code/datums/components/gunpoint.dm

```diff
@@ -0,0 +1,103 @@
+#define GUNPOINT_SHOOTER_STRAY_RANGE 3
+#define GUNPOINT_DELAY_STAGE_2 20
+#define GUNPOINT_DELAY_STAGE_3 80 // cumulative with past stages, so 100 deciseconds
+#define GUNPOINT_MULT_STAGE_1 1
+#define GUNPOINT_MULT_STAGE_2 2
+#define GUNPOINT_MULT_STAGE_3 2.5
+
+
+/datum/component/gunpoint
+	dupe_mode = COMPONENT_DUPE_ALLOWED
+
+	var/mob/living/shooter
+	var/mob/living/target
+	var/obj/item/gun/weapon
+
+	var/stage = 1
+	var/datum/status_effect/status_hold_up
+	var/datum/status_effect/status_held_up
+	var/damage_mult = GUNPOINT_MULT_STAGE_1
+
+/datum/component/gunpoint/Initialize(mob/living/targ, obj/item/gun/wep)
+	if(!isliving(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	shooter = parent
+	target = targ
+	weapon = wep
+	RegisterSignal(targ, list(COMSIG_MOB_ATTACK_HAND, COMSIG_MOB_ITEM_ATTACK, COMSIG_MOB_ATTACK_RANGED, COMSIG_MOVABLE_MOVED), .proc/nap_violation)
+	RegisterSignal(weapon, list(COMSIG_ITEM_DROPPED, COMSIG_ITEM_EQUIPPED), .proc/cancel)
+
+	shooter.visible_message("<span class='danger'>[shooter] aims [weapon] point blank at [target]!</span>", \
+		"<span class='danger'>You aim [weapon] point blank at [target]!</span>", target)
+	to_chat(target, "<span class='userdanger'>[shooter] aims [weapon] point blank at you!</span>")
+	target.emote("gasp") // *extremely bad russian accent* no!
+	status_hold_up = shooter.apply_status_effect(STATUS_EFFECT_HOLDUP)
+	status_held_up = target.apply_status_effect(STATUS_EFFECT_HELDUP)
+
+	addtimer(CALLBACK(src, .proc/update_stage, 2), GUNPOINT_DELAY_STAGE_2)
+
+/datum/component/gunpoint/Destroy(force, silent)
+	qdel(status_hold_up)
+	qdel(status_held_up)
+	return ..()
+
+/datum/component/gunpoint/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_MOVABLE_MOVED, .proc/check_deescalate)
+	RegisterSignal(parent, COMSIG_MOB_APPLY_DAMGE, .proc/flinch)
+
+/datum/component/gunpoint/UnregisterFromParent()
+	UnregisterSignal(parent, COMSIG_MOVABLE_MOVED)
+	UnregisterSignal(parent, COMSIG_MOB_APPLY_DAMGE)
+
+// if you're gonna try to break away from a holdup, better to do it right away
+/datum/component/gunpoint/proc/update_stage(new_stage)
+	stage = new_stage
+	if(stage == 2)
+		to_chat(parent, "<span class='danger'>You steady [weapon] on [target].</span>")
+		to_chat(target, "<span class='userdanger'>[parent] has steadied [weapon] on you!</span>")
+		damage_mult = GUNPOINT_MULT_STAGE_2
+		addtimer(CALLBACK(src, .proc/update_stage, 3), GUNPOINT_DELAY_STAGE_3)
+	else if(stage == 3)
+		to_chat(parent, "<span class='danger'>You have fully steadied [weapon] on [target].</span>")
+		to_chat(target, "<span class='userdanger'>[parent] has fully steadied [weapon] on you!</span>")
+		damage_mult = GUNPOINT_MULT_STAGE_3
+
+/datum/component/gunpoint/proc/check_deescalate()
+	if(!can_see(shooter, target, GUNPOINT_SHOOTER_STRAY_RANGE))
+		cancel()
+
+/datum/component/gunpoint/proc/nap_violation()
```

Comment:
stop making shitty proc names that people have to dig for to understand what they mean, goofball 2.0

Reply:
`/datum/component/gunpoint/proc/ripmacmiller()`